feat: add partial support for p2tr script types

sangaman opened the pull request:
This adds partial support for the `p2tr` script type to `Codes` and `Dimensions`.

The remaining work needed for full support is calculation of input sizes for p2tr multisig script path spends, which can vary depending on which of the three scripts are used in a scenario where a 2 of 3 multisignature scheme is split into a tap tree with three 2 of 2 scripts for each combination. For testing purposes this will require signing/spending p2tr support which will be available in a future version of `@BitGo/utxo-lib`.

This PR also updates some outdated dev dependencies, most significantly moving from `bitgo-utxo-lib` to `@BitGo/utxo-lib` which is needed for testing `p2tr`.

Another follow up to consider would be refactoring the deprecated `TransactionBuilder` calls out of the test suite in favor of the `Psbt` class. Currently the tests print `DEPRECATED: TransactionBuilder sign method arguments will change in v6, please use the TxbSignArg interface` warnings. I tried doing that in this PR but it wasn't simple to do and it's somewhat out of scope for this PR.

Ticket: BG-37329

OttoAllmendinger commented on test/signedTx/txCombinations.ts:
```diff
@@ -1,7 +1,7 @@
 
-// @ts-ignore
-import * as HDKey from 'hdkey';
+import * as bitcoin from '@bitgo/utxo-lib';
+import * as bip32 from 'bip32';
```
:+1:

OttoAllmendinger commented on src/dimensions.ts:
```diff
@@ -106,6 +106,18 @@ export const VirtualSizes = Object.freeze({
+  //   - 4 vbyte nSequence
+  //   - 1 WU witness item count
+  //   - 1 WU witness item size
+  //   - 64 WU schnorr signature
```
you are mixing weight units and vbytes here (1 vbyte = 4 WU)

OttoAllmendinger replied:
ok you actually do the division further down but it still is a bit confusing

OttoAllmendinger commented on src/dimensions.ts:
```diff
@@ -106,6 +106,18 @@ export const VirtualSizes = Object.freeze({
+  //   - 64 WU schnorr signature
+  // 32+4+1+4+(1+1+64)/4 = 57.5 vB - rounded up to 58
+  // See: https://murchandamus.medium.com/2-of-3-multisig-inputs-using-pay-to-taproot-d5faf2312ba3
+  txP2trInputSize: 58,
```
we should call this `txP2trKeypathInputSize`

sangaman replied:
Yeah that makes sense, this is something of a placeholder pending the script path spend logic but might as well call it this now. I'm thinking that eventually we should have this, a size for a first level taptree script spend, and a size for a second level taptree script spend.

OttoAllmendinger commented on src/dimensions.ts:
```diff
@@ -377,6 +397,7 @@ Dimensions.fromInput = function({ index, script, witness }: IOutput, params = {}
+  const p2trInput = Dimensions.sum({ nP2trInputs: 1 });
 
   if (!script.length) {
     if (witness.length > 0) {
```
this branch is now invalid - we cannot assume p2wsh anymore

OttoAllmendinger reviewed (changes requested):
I will submit a commit with some suggestions

sangaman commented:
> The code readability suffers a bit from being a weird tcomb and typescript hybrid.
> 
> We should probably remove the tcomb parts and use vanilla typescript (sacrificing some runtime checks)

I agree and thought this too, but didn't want to make this PR too large since for now I'm just trying to get the basic support for taproot codes for use in WP.

I think this refactoring would make sense as part of migrating this over to BitGoJS if we go with that plan.

sangaman commented:
`npm audit` was failing due to https://github.com/advisories/GHSA-93q8-gq69-wqmw - so I fixed the failure and squashed otto's fix up commit. I'll deal with the script path spend input sizes in a follow-up PR if that's OK.

brandonblack commented on src/dimensions.ts:
```diff
@@ -330,13 +360,14 @@ const mapDimensions = (dim: IDimensions, func: MapFunc) => {
     _.fromPairs(_.map(Dimensions.meta.props, (prop, key) =>
       [key, func((dim as any)[key], key as keyof IDimensions, prop as DimPropertyConstructor)],
-    )) as IDimensions,
+    )) as any,
```
The types here are so confusing, but after some consideration I mostly see why.
